helper/builtinplugins/registry.go

```diff
@@ -0,0 +1,168 @@
+	logicalRabbit "github.com/hashicorp/vault/builtin/logical/rabbitmq"
+	logicalSsh "github.com/hashicorp/vault/builtin/logical/ssh"
+	logicalTotp "github.com/hashicorp/vault/builtin/logical/totp"
+	logicalTransit "github.com/hashicorp/vault/builtin/logical/transit"
+)
+
+// Registry is inherently thread-safe because it's immutable.
+// Thus, rather than creating multiple instances of it, we only need one.
+var Registry = newRegistry()
+
+func newRegistry() *registry {
+	return &registry{
+		credentialBackends: map[string]logical.Factory{
+			"alicloud":   credAliCloud.Factory,
+			"app-id":     credAppId.Factory,
+			"approle":    credAppRole.Factory,
+			"aws":        credAws.Factory,
+			"azure":      credAzure.Factory,
+			"centrify":   credCentrify.Factory,
+			"cert":       credCert.Factory,
+			"gcp":        credGcp.Factory,
+			"github":     credGitHub.Factory,
+			"jwt":        credJWT.Factory,
+			"kubernetes": credKube.Factory,
+			"ldap":       credLdap.Factory,
+			"okta":       credOkta.Factory,
+			"radius":     credRadius.Factory,
+			"userpass":   credUserpass.Factory,
+		},
+		databasePlugins: map[string]func() (interface{}, error){
+			// These four plugins all use the same mysql implementation but with
+			// different username settings passed by the constructor.
+			"mysql-database-plugin":        dbMysql.New(dbMysql.MetadataLen, dbMysql.MetadataLen, dbMysql.UsernameLen),
+			"mysql-aurora-database-plugin": dbMysql.New(credsutil.NoneLength, dbMysql.LegacyMetadataLen, dbMysql.LegacyUsernameLen),
+			"mysql-rds-database-plugin":    dbMysql.New(credsutil.NoneLength, dbMysql.LegacyMetadataLen, dbMysql.LegacyUsernameLen),
+			"mysql-legacy-database-plugin": dbMysql.New(credsutil.NoneLength, dbMysql.LegacyMetadataLen, dbMysql.LegacyUsernameLen),
+
+			"postgresql-database-plugin": dbPostgres.New,
+			"mssql-database-plugin":      dbMssql.New,
+			"cassandra-database-plugin":  dbCass.New,
+			"mongodb-database-plugin":    dbMongo.New,
+			"hana-database-plugin":       dbHana.New,
+		},
+		logicalBackends: map[string]logical.Factory{
+			"ad":         logicalAd.Factory,
+			"alicloud":   logicalAlicloud.Factory,
+			"aws":        logicalAws.Factory,
+			"azure":      logicalAzure.Factory,
+			"cassandra":  logicalCass.Factory,
+			"consul":     logicalConsul.Factory,
+			"gcp":        logicalGcp.Factory,
+			"gcpkms":     logicalGcpKms.Factory,
+			"mongodb":    logicalMongo.Factory,
+			"mssql":      logicalMssql.Factory,
+			"mysql":      logicalMysql.Factory,
+			"nomad":      logicalNomad.Factory,
+			"pki":        logicalPki.Factory,
+			"postgresql": logicalPostgres.Factory,
+			"rabbitmq":   logicalRabbit.Factory,
+			"ssh":        logicalSsh.Factory,
+			"totp":       logicalTotp.Factory,
+			"transit":    logicalTransit.Factory,
+		},
+	}
+}
+
+type registry struct {
+	credentialBackends map[string]logical.Factory
+	databasePlugins    map[string]func() (interface{}, error)
```

Comment:
Might be nice to keep `type BuiltinFactory func() (interface{}, error)` from the now-deleted `builtin.go` and have that be `map[string]BuiltinFactory` in here

Reply:
We call `BuiltinFactory()` in `logical/plugin/plugin.go` so it's definitely needed.

Reply:
Sure! Re-added here: https://github.com/hashicorp/vault/pull/5536/commits/75548176c92d1201f72c90dcc21315d0f50c5118